{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "19248c36",
   "metadata": {},
   "source": [
    "## The analysis pipeline is wrapped into a Python package: probacseq_toolkit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8e6210f",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import os\n",
    "import probacseq_toolkit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06d3ee6",
   "metadata": {},
   "source": [
    "## 1. Map probes to the E.coli reference genome\n",
    "\n",
    "1. Map probes to the E.coli reference with BWA software\n",
    "2. Identify and merge overlapping probes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "57ca1a2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.05 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 1.30 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.03 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.03 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.58 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /data/programs/bwa-0.7.17/bwa index /homes8/jingyuan/Projects/bacteria_sc_multiplex/data/E_coli_reference/Escherichia_coli_w_gca_000184185.ASM18418v1.dna.chromosome.Chromosome.fa\n",
      "[main] Real time: 2.439 sec; CPU: 1.998 sec\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 19550 sequences (1160650 bp)...\n",
      "[M::mem_process_seqs] Processed 19550 reads in 0.451 CPU sec, 0.451 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: /data/programs/bwa-0.7.17/bwa mem ./data/E_coli_reference/Escherichia_coli_w_gca_000184185.ASM18418v1.dna.chromosome.Chromosome.fa ./data/E_coli_reference/probe.fa\n",
      "[main] Real time: 0.540 sec; CPU: 0.488 sec\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./data/mapped_probes.csv\n",
      "CPU times: user 19 s, sys: 126 ms, total: 19.1 s\n",
      "Wall time: 22.3 s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>probe</th>\n",
       "      <th>gene</th>\n",
       "      <th>strand</th>\n",
       "      <th>position</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aaeA_1</td>\n",
       "      <td>aaeA</td>\n",
       "      <td>-</td>\n",
       "      <td>3601479</td>\n",
       "      <td>CAGCTTAGTTTCTTCCATATAGGCCAGTACATAGAAGGAGTTCTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aaeA_2</td>\n",
       "      <td>aaeA</td>\n",
       "      <td>-</td>\n",
       "      <td>3601861</td>\n",
       "      <td>AGTATCTGTCCTTTTTTCACCAGCTGGTTATCATGAACATTCACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>aaeA_3</td>\n",
       "      <td>aaeA</td>\n",
       "      <td>-</td>\n",
       "      <td>3602022</td>\n",
       "      <td>GGCCAGAATGACTAATACGACCGTGATGGCCGTACGGGAGAATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>aaeB_1</td>\n",
       "      <td>aaeB</td>\n",
       "      <td>-</td>\n",
       "      <td>3599466</td>\n",
       "      <td>GAAATTTCGGCAAATCCCCTGGGAACTTATTCATCAGCAAAAACAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>aaeB_2</td>\n",
       "      <td>aaeB</td>\n",
       "      <td>-</td>\n",
       "      <td>3599701</td>\n",
       "      <td>AAACTGACTGAAATGGAAAGTCATCGGGTTATCCAGCACGATAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zur_1</td>\n",
       "      <td>zur</td>\n",
       "      <td>-</td>\n",
       "      <td>4525442</td>\n",
       "      <td>CTGATCGAACAGATGACAGAGCACATAACTGTTGGTGGATTCCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zwf_1</td>\n",
       "      <td>zwf</td>\n",
       "      <td>-</td>\n",
       "      <td>2083199</td>\n",
       "      <td>AGATGCGTCTGATTAAAGGTTTCTGAATAGCTCAGATCCAGCTTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zwf_2</td>\n",
       "      <td>zwf</td>\n",
       "      <td>-</td>\n",
       "      <td>2083369</td>\n",
       "      <td>AAACAGATTCAGTTCAGGTGTTTTGAAATAGACCACGACTTCAGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zwf_3</td>\n",
       "      <td>zwf</td>\n",
       "      <td>-</td>\n",
       "      <td>2083897</td>\n",
       "      <td>CACCGTTTCTTTACCAAGATAGTGGTCGATACGGTAAACCTGGCAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zwf_4</td>\n",
       "      <td>zwf</td>\n",
       "      <td>-</td>\n",
       "      <td>2084160</td>\n",
       "      <td>CAGTGTCATTGACATCGAGATTACAAAAATCCAGACGTGCACTCAG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8999 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     probe  gene strand  position  \\\n",
       "0   aaeA_1  aaeA      -   3601479   \n",
       "1   aaeA_2  aaeA      -   3601861   \n",
       "2   aaeA_3  aaeA      -   3602022   \n",
       "0   aaeB_1  aaeB      -   3599466   \n",
       "1   aaeB_2  aaeB      -   3599701   \n",
       "..     ...   ...    ...       ...   \n",
       "0    zur_1   zur      -   4525442   \n",
       "0    zwf_1   zwf      -   2083199   \n",
       "1    zwf_2   zwf      -   2083369   \n",
       "2    zwf_3   zwf      -   2083897   \n",
       "3    zwf_4   zwf      -   2084160   \n",
       "\n",
       "                                             sequence  \n",
       "0   CAGCTTAGTTTCTTCCATATAGGCCAGTACATAGAAGGAGTTCTGT...  \n",
       "1   AGTATCTGTCCTTTTTTCACCAGCTGGTTATCATGAACATTCACCT...  \n",
       "2   GGCCAGAATGACTAATACGACCGTGATGGCCGTACGGGAGAATTTT...  \n",
       "0   GAAATTTCGGCAAATCCCCTGGGAACTTATTCATCAGCAAAAACAG...  \n",
       "1   AAACTGACTGAAATGGAAAGTCATCGGGTTATCCAGCACGATAATA...  \n",
       "..                                                ...  \n",
       "0   CTGATCGAACAGATGACAGAGCACATAACTGTTGGTGGATTCCACC...  \n",
       "0   AGATGCGTCTGATTAAAGGTTTCTGAATAGCTCAGATCCAGCTTGG...  \n",
       "1   AAACAGATTCAGTTCAGGTGTTTTGAAATAGACCACGACTTCAGAA...  \n",
       "2   CACCGTTTCTTTACCAAGATAGTGGTCGATACGGTAAACCTGGCAC...  \n",
       "3   CAGTGTCATTGACATCGAGATTACAAAAATCCAGACGTGCACTCAG...  \n",
       "\n",
       "[8999 rows x 5 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "probe_file = \"./data/probes_sequence.csv\"\n",
    "bwa_path = \"/data/programs/bwa-0.7.17/bwa\"\n",
    "reference_genome = \"./data/E_coli_reference/Escherichia_coli_w_gca_000184185.ASM18418v1.dna.chromosome.Chromosome.fa\"\n",
    "\n",
    "probe_df = probacseq_toolkit.map_probes(probe_file, bwa_path, reference_genome, merge_overlapping_probes=True)\n",
    "probe_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5edb557",
   "metadata": {},
   "source": [
    "## 2. Make a custom reference genome by stitching the probes together.\n",
    "\n",
    "1. Sort the probes by their position on the reference E.coli\n",
    "2. Convert the probes to its reverse complement if they are on the reverse strand\n",
    "3. Add NNNN paddings to the front and end of each probe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "46a79256",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 9.06 s, sys: 12.7 ms, total: 9.07 s\n",
      "Wall time: 9.18 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "probe_ref_dir = \"./data/probe_reference/\"\n",
    "output_file_prefix = \"E.coli\"\n",
    "probacseq_toolkit.make_fa_and_gtf_reference(probe_df, probe_ref_dir, output_file_prefix,\n",
    "                                            fwd_primer = \"NNNNNNN\", rev_primer = \"NNNNNNN\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80524f3c",
   "metadata": {},
   "source": [
    "## 3. Make 10x reference index with cell ranger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "53de0703",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/programs/cellranger-7.1.0/bin/cellranger mkref --genome=E.coli_probes --fasta=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli.fasta --genes=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli.gtf\n",
      "['/data/programs/cellranger-7.1.0/bin/rna/mkref', '--genome=E.coli_probes', '--fasta=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli.fasta', '--genes=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli.gtf']\n",
      "CPU times: user 3.32 ms, sys: 11.9 ms, total: 15.2 ms\n",
      "Wall time: 1.36 s\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Destination reference folder already exists: /homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes\n",
      "Please delete and start again.\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "path_to_cell_ranger = \"/data/programs/cellranger-7.1.0/bin/cellranger\"\n",
    "genome_name = \"E.coli_probes\"\n",
    "fa_gtf_prefix = \"%s%s\"%(probe_ref_dir,output_file_prefix)\n",
    "output_dir = probe_ref_dir\n",
    "probacseq_toolkit.build_reference_index_with_cell_ranger(path_to_cell_ranger, genome_name,\\\n",
    "                                            fa_gtf_prefix, output_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "888517a0",
   "metadata": {},
   "source": [
    "## 4. Segment reads to each component\n",
    "1. Identify each component with regular expression. We'll need a csv file.\n",
    "2. Merge probe UMI with 10X UMI\n",
    "3. Demux the data by sample_identifier. \n",
    "4. The unidentified fastq file contains read that are either: (1) anchors can't be aligned to the read. (2) Sample_identifier is not in the white list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "722d060f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>segment_name</th>\n",
       "      <th>length</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>anchor</td>\n",
       "      <td>22</td>\n",
       "      <td>CGCTAGTATGGATAGCCGTGTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sample</td>\n",
       "      <td>4</td>\n",
       "      <td>AAGG|GGAA|TTCC|CCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>umi</td>\n",
       "      <td>12</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>anchor</td>\n",
       "      <td>15</td>\n",
       "      <td>GACCCGGTCCATACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>probe</td>\n",
       "      <td>30-70</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>anchor</td>\n",
       "      <td>19</td>\n",
       "      <td>CATAGTTTCTTCGAGCAAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  segment_name length                sequence\n",
       "0       anchor     22  CGCTAGTATGGATAGCCGTGTC\n",
       "1       sample      4     AAGG|GGAA|TTCC|CCTT\n",
       "2          umi     12                       .\n",
       "3       anchor     15         GACCCGGTCCATACA\n",
       "4        probe  30-70                       .\n",
       "5       anchor     19     CATAGTTTCTTCGAGCAAG"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "construct_df = pd.read_csv(\"./data/construct.csv\")\n",
    "construct_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7cc3986",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'(?P<anchor>CGCTAGTATGGATAGCCGTGTC{i<=1,d<=1,s<=3,e<=3})(?P<sample>(AAGG|GGAA|TTCC|CCTT))(?P<umi>.{12})(?P<anchor>GACCCGGTCCATACA{i<=1,d<=1,s<=3,e<=3})(?P<probe>.{30,70})(?P<anchor>CATAGTTTCTTCGAGCAAG{i<=1,d<=1,s<=3,e<=3})'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "construct_regex = probacseq_toolkit.make_construct_regex_pattern(construct_df, insertion=1, deletion=1,\n",
    "                                 substitution=3, error=3)\n",
    "construct_regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f38c4de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 27.3 s, sys: 99.8 ms, total: 27.4 s\n",
      "Wall time: 28.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "r1_fastq = \"./data/30-844302426/00_fastq/SBS_R1_001.fastq.gz\"\n",
    "r2_fastq = \"./data/30-844302426/00_fastq/SBS_R2_001.fastq.gz\"\n",
    "segment_fq_dir = \"./data/30-844302426/segmented_fastq/\"\n",
    "file_prefix = \"SBS\"\n",
    "sample_identifiers = [\"AAGG\",\"GGAA\",\"TTCC\",\"CCTT\"]\n",
    "\n",
    "probacseq_toolkit.segment_fastq_multiplex(construct_regex, r1_fastq, r2_fastq, use_reverse_complement=True,\n",
    "                            sample_identifiers=sample_identifiers, output_dir=segment_fq_dir, file_prefix=file_prefix)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce243447",
   "metadata": {},
   "source": [
    "## 5. Align reads to the reference genome with cell ranger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "551f29ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is the cell ranger command:\n",
      "/data/programs/cellranger-7.1.0/bin/cellranger count --id=AAGG --chemistry=threeprime --transcriptome=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes --fastqs=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/30-844302426/segmented_fastq/AAGG --localcores=6 --localmem=32\n",
      "Martian Runtime - v4.0.10\n",
      "2023-05-10 23:06:50 [runtime] Reattaching in local mode.\n",
      "Serving UI at http://lynx:41145?auth=qUP-jl2FvxhRIkWAy0BkJHXJ8HwPaK1g3whe7Qm6xek\n",
      "\n",
      "\n",
      "Outputs:\n",
      "- Run summary HTML:                         /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/web_summary.html\n",
      "- Run summary CSV:                          /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/metrics_summary.csv\n",
      "- BAM:                                      /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/possorted_genome_bam.bam\n",
      "- BAM BAI index:                            /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/possorted_genome_bam.bam.bai\n",
      "- BAM CSI index:                            null\n",
      "- Filtered feature-barcode matrices MEX:    /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/filtered_feature_bc_matrix\n",
      "- Filtered feature-barcode matrices HDF5:   /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/filtered_feature_bc_matrix.h5\n",
      "- Unfiltered feature-barcode matrices MEX:  /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/raw_feature_bc_matrix\n",
      "- Unfiltered feature-barcode matrices HDF5: /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/raw_feature_bc_matrix.h5\n",
      "- Secondary analysis output CSV:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/analysis\n",
      "- Per-molecule read information:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/molecule_info.h5\n",
      "- CRISPR-specific analysis:                 null\n",
      "- Antibody aggregate barcodes:              null\n",
      "- Loupe Browser file:                       /homes8/jingyuan/Projects/bacteria_sc_multiplex/AAGG/outs/cloupe.cloupe\n",
      "- Feature Reference:                        null\n",
      "- Target Panel File:                        null\n",
      "- Probe Set File:                           null\n",
      "\n",
      "Waiting 6 seconds for UI to do final refresh.\n",
      "Pipestance completed successfully!\n",
      "\n",
      "2023-05-10 23:06:56 Shutting down.\n",
      "Saving pipestance info to \"AAGG/AAGG.mri.tgz\"\n",
      "This is the cell ranger command:\n",
      "/data/programs/cellranger-7.1.0/bin/cellranger count --id=GGAA --chemistry=threeprime --transcriptome=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes --fastqs=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/30-844302426/segmented_fastq/GGAA --localcores=6 --localmem=32\n",
      "Martian Runtime - v4.0.10\n",
      "2023-05-10 23:07:00 [runtime] Reattaching in local mode.\n",
      "Serving UI at http://lynx:42177?auth=0f1cNddBii28Y4qJLQZyupUKen80kjoTK4HVOYuD7kI\n",
      "\n",
      "\n",
      "Outputs:\n",
      "- Run summary HTML:                         /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/web_summary.html\n",
      "- Run summary CSV:                          /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/metrics_summary.csv\n",
      "- BAM:                                      /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/possorted_genome_bam.bam\n",
      "- BAM BAI index:                            /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/possorted_genome_bam.bam.bai\n",
      "- BAM CSI index:                            null\n",
      "- Filtered feature-barcode matrices MEX:    /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/filtered_feature_bc_matrix\n",
      "- Filtered feature-barcode matrices HDF5:   /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/filtered_feature_bc_matrix.h5\n",
      "- Unfiltered feature-barcode matrices MEX:  /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/raw_feature_bc_matrix\n",
      "- Unfiltered feature-barcode matrices HDF5: /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/raw_feature_bc_matrix.h5\n",
      "- Secondary analysis output CSV:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/analysis\n",
      "- Per-molecule read information:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/molecule_info.h5\n",
      "- CRISPR-specific analysis:                 null\n",
      "- Antibody aggregate barcodes:              null\n",
      "- Loupe Browser file:                       /homes8/jingyuan/Projects/bacteria_sc_multiplex/GGAA/outs/cloupe.cloupe\n",
      "- Feature Reference:                        null\n",
      "- Target Panel File:                        null\n",
      "- Probe Set File:                           null\n",
      "\n",
      "Waiting 6 seconds for UI to do final refresh.\n",
      "Pipestance completed successfully!\n",
      "\n",
      "2023-05-10 23:07:07 Shutting down.\n",
      "Saving pipestance info to \"GGAA/GGAA.mri.tgz\"\n",
      "This is the cell ranger command:\n",
      "/data/programs/cellranger-7.1.0/bin/cellranger count --id=TTCC --chemistry=threeprime --transcriptome=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes --fastqs=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/30-844302426/segmented_fastq/TTCC --localcores=6 --localmem=32\n",
      "Martian Runtime - v4.0.10\n",
      "2023-05-10 23:07:10 [runtime] Reattaching in local mode.\n",
      "Serving UI at http://lynx:36841?auth=SqwlkH-rgLGw_3C1GLc8F9JoB5j41ChfMBhubPGV9bc\n",
      "\n",
      "\n",
      "Outputs:\n",
      "- Run summary HTML:                         /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/web_summary.html\n",
      "- Run summary CSV:                          /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/metrics_summary.csv\n",
      "- BAM:                                      /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/possorted_genome_bam.bam\n",
      "- BAM BAI index:                            /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/possorted_genome_bam.bam.bai\n",
      "- BAM CSI index:                            null\n",
      "- Filtered feature-barcode matrices MEX:    /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/filtered_feature_bc_matrix\n",
      "- Filtered feature-barcode matrices HDF5:   /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/filtered_feature_bc_matrix.h5\n",
      "- Unfiltered feature-barcode matrices MEX:  /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/raw_feature_bc_matrix\n",
      "- Unfiltered feature-barcode matrices HDF5: /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/raw_feature_bc_matrix.h5\n",
      "- Secondary analysis output CSV:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/analysis\n",
      "- Per-molecule read information:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/molecule_info.h5\n",
      "- CRISPR-specific analysis:                 null\n",
      "- Antibody aggregate barcodes:              null\n",
      "- Loupe Browser file:                       /homes8/jingyuan/Projects/bacteria_sc_multiplex/TTCC/outs/cloupe.cloupe\n",
      "- Feature Reference:                        null\n",
      "- Target Panel File:                        null\n",
      "- Probe Set File:                           null\n",
      "\n",
      "Waiting 6 seconds for UI to do final refresh.\n",
      "Pipestance completed successfully!\n",
      "\n",
      "2023-05-10 23:07:17 Shutting down.\n",
      "Saving pipestance info to \"TTCC/TTCC.mri.tgz\"\n",
      "This is the cell ranger command:\n",
      "/data/programs/cellranger-7.1.0/bin/cellranger count --id=CCTT --chemistry=threeprime --transcriptome=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes --fastqs=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/30-844302426/segmented_fastq/CCTT --localcores=6 --localmem=32\n",
      "Martian Runtime - v4.0.10\n",
      "2023-05-10 23:07:20 [runtime] Reattaching in local mode.\n",
      "Serving UI at http://lynx:45613?auth=3Pbx3IS3i5Y25rF3Q2zUYTQqTp-kneTBiaa8NEw7EiY\n",
      "\n",
      "\n",
      "Outputs:\n",
      "- Run summary HTML:                         /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/web_summary.html\n",
      "- Run summary CSV:                          /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/metrics_summary.csv\n",
      "- BAM:                                      /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/possorted_genome_bam.bam\n",
      "- BAM BAI index:                            /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/possorted_genome_bam.bam.bai\n",
      "- BAM CSI index:                            null\n",
      "- Filtered feature-barcode matrices MEX:    /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/filtered_feature_bc_matrix\n",
      "- Filtered feature-barcode matrices HDF5:   /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/filtered_feature_bc_matrix.h5\n",
      "- Unfiltered feature-barcode matrices MEX:  /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/raw_feature_bc_matrix\n",
      "- Unfiltered feature-barcode matrices HDF5: /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/raw_feature_bc_matrix.h5\n",
      "- Secondary analysis output CSV:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/analysis\n",
      "- Per-molecule read information:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/molecule_info.h5\n",
      "- CRISPR-specific analysis:                 null\n",
      "- Antibody aggregate barcodes:              null\n",
      "- Loupe Browser file:                       /homes8/jingyuan/Projects/bacteria_sc_multiplex/CCTT/outs/cloupe.cloupe\n",
      "- Feature Reference:                        null\n",
      "- Target Panel File:                        null\n",
      "- Probe Set File:                           null\n",
      "\n",
      "Waiting 6 seconds for UI to do final refresh.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pipestance completed successfully!\n",
      "\n",
      "2023-05-10 23:07:27 Shutting down.\n",
      "Saving pipestance info to \"CCTT/CCTT.mri.tgz\"\n",
      "This is the cell ranger command:\n",
      "/data/programs/cellranger-7.1.0/bin/cellranger count --id=unidentified --chemistry=threeprime --transcriptome=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes --fastqs=/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/30-844302426/segmented_fastq/unidentified --localcores=6 --localmem=32\n",
      "Martian Runtime - v4.0.10\n",
      "2023-05-10 23:07:30 [runtime] Reattaching in local mode.\n",
      "Serving UI at http://lynx:46079?auth=MRF_0aLMRZEGEnVDFT5uWW84koqLFVcswuFAkZHoZR8\n",
      "\n",
      "\n",
      "Outputs:\n",
      "- Run summary HTML:                         /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/web_summary.html\n",
      "- Run summary CSV:                          /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/metrics_summary.csv\n",
      "- BAM:                                      /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/possorted_genome_bam.bam\n",
      "- BAM BAI index:                            /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/possorted_genome_bam.bam.bai\n",
      "- BAM CSI index:                            null\n",
      "- Filtered feature-barcode matrices MEX:    /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/filtered_feature_bc_matrix\n",
      "- Filtered feature-barcode matrices HDF5:   /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/filtered_feature_bc_matrix.h5\n",
      "- Unfiltered feature-barcode matrices MEX:  /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/raw_feature_bc_matrix\n",
      "- Unfiltered feature-barcode matrices HDF5: /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/raw_feature_bc_matrix.h5\n",
      "- Secondary analysis output CSV:            null\n",
      "- Per-molecule read information:            /homes8/jingyuan/Projects/bacteria_sc_multiplex/unidentified/outs/molecule_info.h5\n",
      "- CRISPR-specific analysis:                 null\n",
      "- Antibody aggregate barcodes:              null\n",
      "- Loupe Browser file:                       null\n",
      "- Feature Reference:                        null\n",
      "- Target Panel File:                        null\n",
      "- Probe Set File:                           null\n",
      "\n",
      "Waiting 6 seconds for UI to do final refresh.\n",
      "Pipestance completed successfully!\n",
      "\n",
      "2023-05-10 23:07:37 Shutting down.\n",
      "Saving pipestance info to \"unidentified/unidentified.mri.tgz\"\n",
      "CPU times: user 40 ms, sys: 53.4 ms, total: 93.4 ms\n",
      "Wall time: 49.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "path_to_cell_ranger = \"/data/programs/cellranger-7.1.0/bin/cellranger\"\n",
    "transcriptome = \"/homes8/jingyuan/Projects/bacteria_sc_multiplex/data/probe_reference/E.coli_probes\"\n",
    "for sample_idx in sample_identifiers:\n",
    "    id = sample_idx\n",
    "    fastq_dir = os.path.abspath(\"%s/%s/\"%(segment_fq_dir,sample_idx))\n",
    "    localcores = 6\n",
    "    localmem = 32\n",
    "    probacseq_toolkit.run_cellranger_alignment(path_to_cell_ranger, id, transcriptome,\n",
    "                             fastq_dir, localcores=localcores, localmem=localmem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da95f1b6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
